### ISSUE TITLE
configure does not work on osx

### ISSUE THREAD

**Issue by dijs** [opencog/link-grammar#216]
```
./configure: line 18517: syntax error near unexpected token `2.0.0,'
./configure: line 18517: `AX_PKG_SWIG(2.0.0, SwigF=yes, SwigF=no)'
```

Any ideas?

**Comment by linas** [comment]
The standard recommendation is that you should download version 5.3.1 and work with that, instead of trying to build what's in github.  Here: http://www.abisource.com/downloads/link-grammar/5.3.1/link-grammar-5.3.1.tar.gz

If you try to build from github sources, you need to install all dependencies; in this case, its the "autoconf archive", which contains a large number of autoconf macros.

**Comment by linas** [comment]
Hey, I just realized you are the node.js maintainer ... could I convince you to merge the node.js code into this repo? I believe that this would simplify things for users, while probably not making maintenance any harder ...

**Comment by dijs** [comment]
That is a great idea actually, that way node users always have the latest
code. Do you push up stable library binaries which the JavaScript could use?

**Comment by linas** [comment]
I don't maintain any pre-compiled binaries; I've got too much to do already.  There are wayy too many different types of OS'es and all their peculiar quirks (for me) to try to do this.

I think there might be some kind of web-services out there that will automatically build binaries for you.  Never looked into it.

The correct solution would be to have every OS have a maintainer doing this. Someone used to do this for Debian, but they dropped out.  There are some irregular updates for Red hat. Unaware of any activity for Mac Ports. And Windows seems like a lost cause.

**Comment by dijs** [comment]
Okay. Let me try to compile and use 5.3.1 for node usage.

**Comment by dijs** [comment]
So 5.3.1 is still a nightmare on my machine. Where is the documentation for compiling the source?

Do I use configure? Or the autogen shell script?

**Comment by linas** [comment]
The README explains how to build the package. Normally, you should never run autogen unless you are familiar with the autotools suite. For regular users, running configure is enough.

If you tell me what OS you are on, what steps you took, and what the error messages are, then maybe I can help. Its simply not supposed to be that hard.

**Comment by dijs** [comment]
OS: OSX Yosemite
Steps: 
- Downloaded 5.3.1 and extracted
- Ran `./configure --disable-java-bindings --disable-editline`
- Configure output: https://gist.github.com/dijs/2ae366f4948177b3fdb6
- Ran `make`
- Got error:
```
print.c:16:10: fatal error: 'malloc.h' file not found
#include <malloc.h>
         ^
1 error generated.
make[2]: *** [print.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```
- Changed configure to `./configure --disable-java-bindings --disable-editline CFLAGS="-I/usr/include/malloc"`
- Ran make again
- Getting error: https://gist.github.com/dijs/4f93be50064c9f326c3c

**Comment by linas** [comment]
Huh. googling this: "OSX malloc.h"  bring up the following:

https://code.google.com/p/word2vec/issues/detail?id=17

which suggests changing malloc.h to stdlib.h  -- could you edit print.c by hand, and change the malloc.h at the top to stdlib.h and let me know how it goes?

I'll spin a new version 5.3.2 tomorrow or in a few days, depending on how this goes.

**Comment by linas** [comment]
For the link errors, a quick search of google reveals these:
http://stackoverflow.com/questions/16352833/linking-with-clang-on-os-x-generates-lots-of-symbol-not-found-errors
https://code.google.com/p/v8/issues/detail?id=3072
https://groups.google.com/forum/#!topic/capnproto/91O56AkDAFY

I'm trying to figure out the work-around now.

**Comment by linas** [comment]
OK, so according to the first URL, mac/clang needs to have `-stdlib=libc++` instead of `-std=c++03`  I beleive this occurs exactly once in configure, could you edit configure, make this change, and try again?

**Comment by linas** [comment]
The second URL suggests that `-std=c++03` should be replaced by `-std=c++11 -stdlib=libc++` which sounds like a better solution ...

**Comment by linas** [comment]
And this URL http://stackoverflow.com/questions/9759279/pass-stdlib-libc-to-c-linker-with-autotools suggests that the linker work-around is too say this:
```
./configure CXXFLAGS="-O2 -g -stdlib=libc++" LDFLAGS="-stdlib=libc++"
```
I am guessing that this third approach is the best yet. Let me know if it works.

**Comment by dijs** [comment]
Okay. So.

I swapped malloc.h to stdlib.h

Then changed the one line in configure to `CXXFLAGS="${CXXFLAGS} -O3 -Wall -stdlib=libc++"`

But after running
`./configure --disable-java-bindings --disable-editline CXXFLAGS="-O2 -g -stdlib=libc++" LDFLAGS="-stdlib=libc++"`

Still getting the same `ld: symbol(s) not found for architecture x86_64` error

**Comment by linas** [comment]
your machine is a 64-bit machine, and you do have the 64-bit c++ libraries installed, right?

Can you compile and run any C++ code, e.g. some "hello world"  test?

**Comment by linas** [comment]
I noticed that you were using gcc instead of clang; normally, on OSX, clang is the compiler, and so that makes me think that maybe parts of the required C++ devel environment are not installed.  Why did you over-ride clang and choose gcc?

**Comment by dijs** [comment]
Wow. I feel stupid. Your right, I must not have the 64bit c++ libs... Hello World doesn't work.

Any idea how to get those?

And I do not remember ever overriding clang with gcc, I do not work in C++, ever, unless it is compiling other third party source code.

**Comment by linas** [comment]
No clue how to get the devel environment; I assume it comes from the Apple Store. 

Do you have the 32-bit environment installed?  LG doesn't have to be 64-bit; it just defaults to that because Java needs to be 64-bit.  You can hack around this by editing configure and removing  "-arch x86_64" from CFLAGS

Its possible that you need 32-bit LG anyway, for javascript? Or maybe it doesn't matter?

(I haven't tested 32-bit LG in a very long time, I assume it still works)

**Comment by dijs** [comment]
I think I finally have this thing figured out.

Got the `make` to by using this `configure`:

`./configure --disable-java-bindings --disable-editline CFLAGS="-lc++ -arch x86_64" CXXFLAGS="-O2 -g -stdlib=libc++" LDFLAGS="-stdlib=libc++"`

from http://stackoverflow.com/questions/7533321/error-when-compiling-some-simple-c-code

Now I have all the object files. I need to build a dylib from the object files now.

**Comment by linas** [comment]
So, make should generate a `liblink-grammar.so.5` somewhere.   I'm guessing its enough to rename it to `liblink-grammar.dylib` or something like that.  Actually, I would expect the linker to automatically call it dylib on the mac, and NOT call it so

**Comment by dijs** [comment]
I do not see either. I see a "liblink-grammar.la", but that is only 1kb in size.

**Comment by linas** [comment]
Look in the build ./link-grammar/.libs directory. The la file is a text file that contains install and libtool runtime directives.  Take a look at it, it might cntain useful info

**Comment by dijs** [comment]
There it is! Awesome. Default on osx hides "dot" directories... I hate that. 

Thanks for the help, I am going to test this against my node module now. And hopefully maintain a few different OS binaries as needed in there.

**Comment by linas** [comment]
Be aware that there have been "minor" changes in the API betwen 4.1 and 5.3 -- fat links were removed,  The technique for opening the dict files was simplified.  One or two other minor things that were useless were removed.  Strings returned by the API now have to be explicitly freed, or there is a memleak.

**Comment by linas** [comment]
I just published a version 5.3.2 wish should build without hiccups or special treatment on OSX.  The website should show this "real soon now", when it updates.

I'm closing; for any new, additional problems, please open a new bug report.
